[api][webui] Drop CacheLine model

bgeuken opened the pull request:
This model was used as a DB representation of a subset of data we cache
with Rails caching mechanism.
There are a couple of reasons why this isn't needed:
* Invalidating caches can be done by using proper cache keys and making
  sure relations use 'touch', etc.
* Having a DB presentation of cached data kind of defeats the prupose of
  caching. It might still be faster than very complex SQL queries, but
  also slower than fetching data from memcache.
* The cache_lins table grew up to 25 million entries. In other words we
  had the same issues here, one has with finding a good caching key.
  That is, making sure cached data get's invalidated properly.
* All that was stored by CacheLine was generated urls to a package /
  project and the result of a package existance call. That doesn't seem
  to require advanced caching mechanisms.

adrianschroeter commented on src/api/app/models/project.rb:
```diff
@@ -778,7 +776,7 @@ def can_be_released_to_project?(target_project)
 
   def exists_package?(name, opts = {})
-    CacheLine.fetch([self, 'exists_package', name, opts], project: self.name, package: name) do
+    Rails.cache.fetch("exists_package-#{cache_key}-#{name}-#{opts}") do
```
the opts hash needs to get expanded here, right?

bgeuken replied:
Ah, good point :+1:

coolo commented on src/api/app/helpers/webui/webui_helper.rb:
```diff
@@ -330,9 +330,10 @@ def project_or_package_link(opts)
+    # only care for database entries
+    prj = Project.where(name: opts[:project]).select(:id, :name).first
+    # Expires in 15 minutes so that changes of local and remote packages eventually result in an update
+    Rails.cache.fetch(['project_or_package_link', prj.cache_key, opts], expires_in: 15.minutes) do
```
expires in 15 minutes is kind of 'fuck yourself' - this optimization was mainly for long request lists where we checked for existance and rendered links all the time. 15 minutes is not worth the caching

bgeuken replied:
Could also be an hour or two. Not a big deal to change that.

adrianschroeter commented:
can you please also drop the former migration which just modifies this table? 
(can also cause problems when running, right?)

bgeuken commented:
@coolo Is it just me or is viewing requests much faster now?

coolo commented:
Hard to say, but they are not overly slaggish right now :)

bgeuken commented:
Updated:
* Rebased on master
* Removed leftovers from events (triggering a cache cleanup job)
* Skipped altering the table done by previous migration

@adrianschroeter I didn't like the idea of removing a migration that might have been run on some systems. I hope it's okay to you to just skip the migration.

evanrolfe commented on src/api/db/migrate/20170317094221_change_cache_lines_key.rb:
```diff
@@ -1,5 +1,8 @@
 class ChangeCacheLinesKey < ActiveRecord::Migration[5.0]
   def up
+    # CacheLine get's dropped with the next migration
+    return
```
Can you just delete this migration instead?

bgeuken replied:
Ok. Although I don't like the idea of changing our migration history

evanrolfe commented on src/api/app/helpers/webui/webui_helper.rb:
```diff
@@ -330,9 +330,10 @@ def project_or_package_link(opts)
-      # only care for database entries
-      prj = Project.where(name: opts[:project]).select(:id, :name).first
+    # only care for database entries
+    prj = Project.where(name: opts[:project]).select(:id, :name).first
```
You are only selecting id and name from projects but you call `prj.cache_key` on line 336? Maybe just get rid of the `select` since I dont think there is really that much performance gain from only selecting the specific columns as opposed to selecting all of them?
```ruby
prj = Project.find_by(name: opts[:project])
```

bgeuken replied:
Autsch. I missed that one:/ Added the updated_at attribute now.

> evanrolfe 17 hours ago
> prj = Project.find_by(name: opts[:project])

Last time I tried, rubocop wanted me to use .where().first

evanrolfe commented on src/api/app/helpers/webui/webui_helper.rb:
```diff
@@ -330,9 +330,10 @@ def project_or_package_link(opts)
+    # only care for database entries
+    prj = Project.where(name: opts[:project]).select(:id, :name).first
+    # Expires in 2 hours so that changes of local and remote packages eventually result in an update
+    Rails.cache.fetch(['project_or_package_link', prj.cache_key, opts], expires_in: 2.hours) do
```
2 hours seems like a long time for this? Doesn't that mean that if the link changes then it will be broken for 2 hours until the cache resets?

coolo replied:
that's why you have a .cache_key that needs updating whenever project or packages change. The 2 hours are just for remote objects - where we have no way to verify

evanrolfe commented on src/api/db/structure.sql:
```diff
@@ -11,7 +11,7 @@ CREATE TABLE `architectures` (
   `available` tinyint(1) DEFAULT '0',
   PRIMARY KEY (`id`),
-  UNIQUE KEY `arch_name_index` (`name`) USING BTREE
+  UNIQUE KEY `arch_name_index` (`name`)
```
Please run `rake db:drop db:create db:migrate` and then commit db/structure.sql so we don't lose all these "USING BTREE" references. This is clarified in the mailing list [email](https://lists.opensuse.org/opensuse-buildservice/2017-04/msg00009.html) I sent out.

bgeuken replied:
There is a reason why this commit was named "update structure.sql for now";-)
I'll change it later.

hennevogel commented:
@bgeuken you should rebase and address the test failures. They look severe

DavidKang commented:
@bgeuken, what is the state of this PR? Should we close it?
